### ISSUE TITLE
There is still a bug in some static variables

### ISSUE THREAD

**Issue by Rolige** [sirbrillig/phpcs-variable-analysis#82]
Hi again,

Regarding to the previous issue #79, now I have found a new bug with static variables, you can reproduce it with this code:

```php
<?php

class MyClass extends MyForms
{
    private static $defaultClass = 'OtherClass';

    final public static function getForm($class = false)
    {
        if ($class !== false) {
            return new $class();
        }

        return new self::$defaultClass();
    }
}
```

Please check this capture.

![image](https://user-images.githubusercontent.com/2866694/56461232-c55be180-6374-11e9-87b7-0c02ba5793a1.png)

**Comment by sirbrillig** [comment]
It looks like this bug was introduced by #27

**Comment by sirbrillig** [comment]
Are you sure this is a bug? I think that actually it's a correct warning. See the section titled "When calling static methods, the function call is stronger than the static property operator" in [this PHP doc](https://php.net/manual/en/functions.variable-functions.php). I actually had to fix this behavior in #27.

In your example, `return new self::$defaultClass();` will first evaluate `$defaultClass` and then look for a static property on `self` which matches that name. However, `$defaultClass` is not defined. I think you might have meant `return new self::defaultClass();`?

**Comment by Rolige** [comment]
Hi @sirbrillig and thanks for response, I was trying to change the code after your comment and I discovered this...

Why does my previous code throw a warning and this does not?
```php
<?php

class MyClass extends MyForms
{
    private static $defaultClass = 'OtherClass';

    final public static function getForm($class = false)
    {
        $defaultClass = self::$defaultClass;

        if ($class !== false) {
            return new $class();
        }

        return new self::$defaultClass();
    }
}
```

**Comment by sirbrillig** [comment]
That example is not an issue, I think, because in that case, `$defaultClass` _is_ defined by the time it's used. PHP is weird this way. 

In the second example, you're creating a local variable called `$defaultClass`, which you're assigning with the value of `self::$defaultClass`, which evaluates to the static property and thus has the value of `'OtherClass'`. Then you're returning `self::$defaultClass()`. Because the function call is stronger than the static reference, essentially that means that those parenthesis evaluate first, before the `self::` part. So `self::$defaultClass()` first evaluates `$defaultClass()`. In the original code, that was causing an error (and also, the sniff was warning you about it) since `$defaultClass` didn't exist, but in the second example, it _does_ exist, so it will be evaluated into `OtherClass()`. That function does not exist (it's a class, not a function), so you'll probably get a PHP error, but it's a different error and one that this sniff cannot detect because it only looks to make sure variables exist, not functions.

I understand what you're trying to do: dynamically access the static property and then use that as a class name. This is tricky to do in PHP, though. I'm not sure if it will work, but maybe try this:

```php
<?php

class MyClass extends MyForms
{
    private static $defaultClass = 'OtherClass';

    final public static function getForm($class = false)
    {
        if ($class !== false) {
            return new $class();
        }

        return new (self::$defaultClass)();
    }
}
```

**Comment by Rolige** [comment]
Thank u so much @sirbrillig, we will review it!
